### ISSUE TITLE
Proposal: Change declaration of `RangeQuery`

### ISSUE THREAD

**Issue by flobernd** [elastic/elasticsearch-specification#2545]
## Problem Description

`RangeQuery` is currently declared as:

```ts
/** @codegen_names date, number, terms */
// Note: deserialization depends on value types
export type RangeQuery = DateRangeQuery | NumberRangeQuery | TermsRangeQuery
```

This declaration is highly problematic for statically typed clients as there is no way to correctly **de**serialize into the correct union variant, because for example the following variants share the same underlying JSON type:

```ts
export class TermsRangeQuery extends RangeQueryBase {
  gt?: string
```

```ts
export class DateRangeQuery extends RangeQueryBase {
  gt?: DateMath // string
```

## Proposal

It seems like on server side, there is only one `RangeQuery` with an `object/any` type and the actual semantics are applied on base of the type mappings.

It would make sense to reflect this in our specification while still providing _hints_ about potential specializations to the code generator.

The naiv approach would be to change the specification like this:

```ts
export class RangeQuery extends RangeQueryBase {
  gt?: any
  // ...
}
```

This would allow serialization and de-serialization of arbitrary value types. However, this would as well complicate building queries in statically typed languages.

A better way could be a construct like this:

```ts
export type RangeQuery = GenericRangeQuery<any>;

export class GenericRangeQuery<T> extends RangeQueryBase {
  gt?: T
   // ...
}
```

Potentially with new annotations:

```ts
@known_generic = [DateMath, "DateRangeQuery" ]
@known_generic = [double  , "NumberRangeQuery"]
@known_generic = [string  , "TermRangeQuery"  ]
export type RangeQuery = GenericRangeQuery<any>;
```

Based on this, generators could generate specialized code only for serialization (deserialization always uses `DateRange<any>`):

```csharp
// Untyped/any variant always used for deserialization; can optionally be used by the user when building queries
public class RangeQuery
{
	public object? Gt { get; set; }
}

public abstract class GenericRangeQuery<T> : RangeQuery
{
	public new T? Gt { get; set; }
}

// Specialized range query for numbers
public sealed class NumberRangeQuery : GenericRangeQuery<double>
{
}

// Specialized range query for dates
public sealed class DateRangeQuery : GenericRangeQuery<DateMath>
{
}
```

## Side Notes

There are more types for which the same issue applies, for example:

```ts
/** @codegen_names date, numeric, geo */
// Note: deserialization depends on value types
export type DecayFunction =
  | DateDecayFunction
  | NumericDecayFunction
  | GeoDecayFunction
```

```ts
/** @codegen_names geo, date */
// Note: deserialization depends on value types
export type DistanceFeatureQuery =
  | GeoDistanceFeatureQuery
  | DateDistanceFeatureQuery
```

cc @elastic/clients-team @swallez @technige

**Comment by swallez** [comment]
> It seems like on server side, there is only one RangeQuery with an object/any type and the actual semantics are applied on base of the type mappings.

That's correct: this is a mappings-aware deserialization. Something we definitely cannot to client-side. `RangeQuery`, as it is defined, is indeed serializable but not deserializable. This was deemed acceptable since it's only found in requests and not in responses (at least when it was introduced).

However, the Java client ended up simplifying it by replacing it with the definition of `DateRangeQuery` with `DateMath` types replaced with `UserDefinedValue` (i.e. the equivalent of `GenericRangeQuery<UserDefinedValue>`)

`DateRangeQuery` was chosen as the blueprint because it has additional `format` and `timezone` fields that are not found in other variants. So the `GenericRangeQuery` would need to have those.

That being said, I don't see how the new annotation solve the issue, since deserialization is still impossible. I may be missing something.

**Comment by flobernd** [comment]
The idea is to implement asymmetric (de-)serialization.

RangeQuery<any> would always be used for deserialization. The user would have to perform type checks on the any/object fields to work with the data. The deserialized object would however still be round-trippable, which is a huge bonus.

The specialized variants are never used for deserialization. They are only meant to simplify serialization/query building. Instead of RangeQuery<any>, the user could pass RangeQuery<DateMath> and work with the DateMath fields in a native way instead of having to deal with any/object.

**Comment by flobernd** [comment]
I created a draft PR that, for now, keeps the current declarations, but slightly rearranges them to always follow this pattern for these kind of untagged unions:

```ts
export class MyTypeBase<T1, T2, ...> { ... }

export class MyTypeSpecialized1 extends MyTypeBase<int> {}
export class MyTypeSpecialized2 extends MyTypeBase<int> {}
export class MyTypeSpecialized3 extends MyTypeBase<int> {}

/** @codegen_names 1, 2, 3*/
// Note: deserialization depends on value types
export type RangeQuery = MyTypeSpecialized1 | MyTypeSpecialized2 | MyTypeSpecialized3 
```

It already was like this for `DistanceFeatureQuery`, but not for `RangeQuery` and `DecayFunction`.

---

I think this is still not perfect as we declare artificial types in the specification that do not really exist on server-side. The annotation based solution is preferrable imo, but for the time being, this change at least allows code generators to detect this specific pattern and drive specialized code generation.

**Comment by flobernd** [comment]
We decided to go with the solution I provided in #2548 with a small addition of adding a new `variant` type `untagged` for these kinds of unions to:

- Make this case easier detectable by the code generators
- To make this construct more explicit for new contributors
- To be able to add a compiler validation of the pattern based on the presence of this new variant decorator (later)

**Comment by flobernd** [comment]
This has been implemented in #7428
